File: src/main/java/hiiragi283/material/tile/TileEntityMachineExtender.java
package hiiragi283.material.tile;

import hiiragi283.material.api.tile.HiiragiTileEntity;
import hiiragi283.material.util.OptionalUtil;
import net.minecraft.block.BlockHorizontal;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.fluids.FluidUtil;
import org.jetbrains.annotations.Nullable;

import javax.annotation.ParametersAreNonnullByDefault;

@ParametersAreNonnullByDefault
public class TileEntityMachineExtender extends HiiragiTileEntity {

    //    Capability    //

    @Override
    public boolean hasCapability(Capability<?> capability, @Nullable EnumFacing facing) {
        return OptionalUtil.getTile(world, getPos().offset(getState().getValue(BlockHorizontal.FACING)))
                .map(tile -> tile.hasCapability(capability, facing))
                .orElse(super.hasCapability(capability, facing));
    }

    @Nullable
    @Override
    public <T> T getCapability(Capability<T> capability, @Nullable EnumFacing facing) {
        return OptionalUtil.getTile(world, getPos().offset(getState().getValue(BlockHorizontal.FACING)))
                .map(tile -> tile.getCapability(capability, facing))
                .orElse(super.getCapability(capability, facing));
    }

    //    Event    //

    @Override
    public boolean onTileActivated(World world, BlockPos pos, EntityPlayer player, EnumHand hand, EnumFacing facing) {
        if (!world.isRemote) {
            return FluidUtil.interactWithFluidHandler(player, hand, world, pos.offset(getState().getValue(BlockHorizontal.FACING)), getState().getValue(BlockHorizontal.FACING).getOpposite());
        }
        return true;
    }

}

File: src/main/kotlin/hiiragi283/material/tile/TileEntityMachineExtender.kt
package hiiragi283.material.tile

import hiiragi283.material.api.tile.HiiragiTileEntity
import hiiragi283.material.util.getTile
import net.minecraft.block.BlockDirectional
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.fluids.FluidUtil

class TileEntityMachineExtender : HiiragiTileEntity() {

    //    Capability    //

    override fun hasCapability(capability: Capability<*>, facing: EnumFacing?): Boolean {
        return getTile<TileEntity>(world, getPos().offset(getState().getValue(BlockDirectional.FACING)))
            ?.takeUnless { it is TileEntityMachineExtender }
            ?.hasCapability(capability, facing)
            ?: super.hasCapability(capability, facing)
    }

    override fun <T> getCapability(capability: Capability<T>, facing: EnumFacing?): T? {
        return getTile<TileEntity>(world, getPos().offset(getState().getValue(BlockDirectional.FACING)))
            ?.takeUnless { it is TileEntityMachineExtender }
            ?.getCapability(capability, facing)
            ?: super.getCapability(capability, facing)
    }

    //    Event    //

    override fun onTileActivated(
        world: World,
        pos: BlockPos,
        player: EntityPlayer,
        hand: EnumHand,
        facing: EnumFacing
    ): Boolean {
        return if (!world.isRemote) {
            FluidUtil.interactWithFluidHandler(
                player,
                hand,
                world,
                pos.offset(getState().getValue(BlockDirectional.FACING)),
                getState().getValue(BlockDirectional.FACING).opposite
            )
        } else true
    }

}